### PR TITLE
Wcf support

### PR DESCRIPTION
**Pull request by danielmarbach**
- [X] Moved out WCF service exposure from host to dedicated WCF solution in snippets
- [X] WCF is a subfolder and no longer part of hosting
- [X] Simplified naming of files and added necessary redirects
- [x] Upgrade guide

### CODE REVIEW

**Comment by danielmarbach** [comment]
@Particular/docs-maintainers can you have an initial glance at this?

**Comment by danielmarbach** [comment]
@indualagarsamy have a look

File: Snippets/Wcf/Wcf_1/Binding/Message.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Wcf1.Binding
+{
+    public class Request :
+        ICommand
+    {
+    }
+
+    public class Response :
+    IMessage
```
Comment (SzymonPobiega) [inline Snippets/Wcf/Wcf_1/Binding/Message.cs]: Inconsistent indentation

File: Snippets/Wcf/Wcf_1/Binding/Usage.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+        void Simple(EndpointConfiguration configuration)
+        {
+            #region WcfOverrideBinding
+
+            var wcfSettings = configuration.Wcf();
+            wcfSettings.Binding(service => {
+                if (service == typeof(MyService))
+                {
+                    return new BindingConfiguration(new NetNamedPipeBinding(), new Uri("net.pipe://localhost/MyService"));
```
Comment (SzymonPobiega) [inline Snippets/Wcf/Wcf_1/Binding/Usage.cs]: I think it would better to break this line

File: Snippets/Wcf/Wcf_1/Binding/Usage.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                }
+
+                if (service == typeof(CancelOrderService))
+                {
+                    return new BindingConfiguration(new BasicHttpBinding(), new Uri("http://localhost:9009/services/cancelOrder"));
```
Comment (SzymonPobiega) [inline Snippets/Wcf/Wcf_1/Binding/Usage.cs]: Same as above, long lines tend to not render well

File: Snippets/Wcf/Wcf_1/Cancellation/Message.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Wcf1.Cancellation
+{
+    public class Request :
+        ICommand
+    {
+    }
+
+    public class Response :
+    IMessage
```
Comment (SzymonPobiega) [inline Snippets/Wcf/Wcf_1/Cancellation/Message.cs]: Inconsistent indentation.

File: Snippets/Wcf/Wcf_1/Cancellation/Usage.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        class MyService : WcfService<Request, Response>
+        {
+        }
+
+        void Simple(EndpointConfiguration configuration)
+        {
+            #region WcfCancelRequest
+
+            var wcfSettings = configuration.Wcf();
+            wcfSettings.CancelAfter(service => service == typeof(MyService) ? TimeSpan.FromSeconds(5) : TimeSpan.FromSeconds(60));
```
Comment (SzymonPobiega) [inline Snippets/Wcf/Wcf_1/Cancellation/Usage.cs]: Break the line

File: Snippets/Wcf/Wcf_1/Enum/Handler.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
+using System.Threading.Tasks;
+using NServiceBus;
+
+namespace Wcf1.Enum
+{
+
```
Comment (SzymonPobiega) [inline Snippets/Wcf/Wcf_1/Enum/Handler.cs]: 🔥

File: Snippets/Wcf/Wcf_1/Enum/Usage.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+﻿using NServiceBus;
+
+namespace Wcf1.Enum
+{
+    class Usage
+    {
+        #region WcfEnum
```
Comment (SzymonPobiega) [inline Snippets/Wcf/Wcf_1/Enum/Usage.cs]: New lines after begin region and before endregion.

File: Snippets/Wcf/Wcf_1/Object/Usage.cs [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    class Usage
+    {
+        #region WcfObjectCallback
+
+        class MyService : WcfService<Message, ResponseMessage>
+        {
+        }
+
+        #endregion
+
```
Comment (SzymonPobiega) [inline Snippets/Wcf/Wcf_1/Object/Usage.cs]: 🔥

File: Snippets/Wcf/Wcf_1/Routing/Message.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+{
+    using NServiceBus;
+
+    public class Request :
+        ICommand
+    {
+    }
+
+    public class Response :
+    IMessage
```
Comment (SzymonPobiega) [inline Snippets/Wcf/Wcf_1/Routing/Message.cs]: Inconsistent indentation

File: Snippets/Wcf/Wcf_1/Routing/Usage.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        class MyService : WcfService<Request, Response>
+        {
+        }
+
+        void Simple(EndpointConfiguration configuration)
+        {
+            #region WcfRouting
+
+            var wcfSettings = configuration.Wcf();
+            wcfSettings.SendOptions(service =>
```
Comment (SzymonPobiega) [inline Snippets/Wcf/Wcf_1/Routing/Usage.cs]: I ❤️ this approach.

File: nservicebus/wcf/index.md [hunk 11]
```diff
@@ -1,22 +1,43 @@
...
 
-Note: Starting from NServiceBus Core Versions 6 and above, the built-in support for WCF integration has been removed. For more details refer to the [Upgrade Guide](/nservicebus/upgrades/host-6to7.md).
+To handle responses on the client, the client (or the sending process) must have its own queue and cannot be configured as a SendOnly endpoint. When messages arrive in this queue, they are handled just like on the server by a message handler:
 
-Inherited from `NServiceBus.WcfService<TCommand, TErrorCode>`, as shown below. `TCommand` is the message type of the request. `TErrorCode` must be an enumerated type, and should represent the result of processing the command. Example:
+snippet:WcfEmptyHandler
+
+## Prerequisites for WCF functionality
+
+In NServiceBus Version 4 and above the WCF support has been moved to [NServiceBus.Host NuGet package](https://www.nuget.org/packages/NServiceBus.Host).
```
Comment (SzymonPobiega) [inline nservicebus/wcf/index.md]: It would be awesome to make these specific version info partials.

Comment (danielmarbach) [inline nservicebus/wcf/index.md]: is this worth it?

Comment (SzymonPobiega) [inline nservicebus/wcf/index.md]: I think it always makes a multi-version page more readable. Hard to navigate all these "In version X and above, in version Y and below"

Comment (SzymonPobiega) [inline nservicebus/wcf/index.md]: OK. You convinced me 👍

**Comment by danielmarbach** [comment]
@Particular/docs-maintainers this is as ready as it can be. I will run a few spell checks. Could you please help me out to get this donedone?

File: nservicebus/wcf/index.md [hunk 12]
```diff
@@ -1,22 +1,43 @@
...
+
+In NServiceBus.Host Version 5 to 6 WCF support is built into the NuGet package.
```
Comment (indualagarsamy) [inline nservicebus/wcf/index.md]: >In NServiceBus.Host Version 5 to 6 WCF support is built into the NuGet package.

Can we combine both this sentence and the next? Suggestion below: 

In Versions 5 and 6 of the NServiceBus.Host, support for WCF integrations was built into the host. Starting from Versions 7 and above, WCF support has been separated into its own nuget package, [NServiceBus.Wcf](https://www.nuget.org/packages/NServiceBus.Wcf). This package also has a dependency on [NServiceBus.Callbacks](https://www.nuget.org/packages/NServiceBus.Callbacks/) and the endpoint hosting the WCF service also needs to [configure the callbacks](/nservicebus/messaging/callbacks.md) appropriately.

File: nservicebus/wcf/index.md [hunk 13]
```diff
@@ -1,22 +1,43 @@
...
+
+In NServiceBus.Host Version 7 and above WCF support is shipped as [NServiceBus.Wcf NuGet package](https://www.nuget.org/packages/NServiceBus.Wcf). The package has a dependency to `NServiceBus.Callback`. The endpoint hosting the WCF services needs to configure the [callbacks accordingly](/nservicebus/messaging/callbacks.md).
```
Comment (indualagarsamy) [inline nservicebus/wcf/index.md]: Combine this paragraph with the previous one (see the above suggestion)

File: nservicebus/wcf/index.md [hunk 14]
```diff
@@ -1,22 +1,43 @@
...
+
+In NServiceBus.Host Version 5 to 6 WCF support is built into the NuGet package.
+
+In NServiceBus.Host Version 7 and above WCF support is shipped as [NServiceBus.Wcf NuGet package](https://www.nuget.org/packages/NServiceBus.Wcf). The package has a dependency to `NServiceBus.Callback`. The endpoint hosting the WCF services needs to configure the [callbacks accordingly](/nservicebus/messaging/callbacks.md). 
+
+For more details refer to the [Upgrade Guide](/nservicebus/upgrades/host-6to7.md).
+
+## Expose a WCF service
+
+Inherited from `NServiceBus.WcfService<TRequest, TResponse>`, as shown below. `TRequest` is the message type of the request. `TResponse` represents the result of processing the command and can be any type that is supported by the `NServiceBus.Callback` package.
```
Comment (indualagarsamy) [inline nservicebus/wcf/index.md]: >Inherited from `NServiceBus.WcfService<TRequest, TResponse>`, as shown below.

How about:
To expose the endpoint as a WCF service, inherit from `NServiceBus.WcfService<TRequest, TResponse>`, as shown below:

File: nservicebus/wcf/index.md [hunk 15]
```diff
@@ -1,22 +1,43 @@
...
+
+NOTE: In previous versions of the WCF support `TResponse` must be an enumerated type. To reply with enumeration types, the replying endpoint needs to reference `NServiceBus.Callback` and [configure](/nservicebus/messaging/callbacks.md) it accordingly.
```
Comment (indualagarsamy) [inline nservicebus/wcf/index.md]: Can you move the note to below the snippet?

File: nservicebus/wcf/index.md [hunk 16]
```diff
@@ -1,22 +1,43 @@
...
+
+Example:
 
 snippet:ExposeWCFService
 
-NOTE: Version 4 and above of NServiceBus `WcfService<TCommand, TErrorCode>` has been moved to [NServiceBus.Host NuGet package](https://www.nuget.org/packages/NServiceBus.Host), so this package needs to be referenced.
+## Configure binding and address of WCF service
 
 And finally expose the WCF service via the config file, for the example above the XML would look something like:
```
Comment (indualagarsamy) [inline nservicebus/wcf/index.md]: >And finally expose the WCF service via the config file, for the example above the XML would look something like:

I understand, this wasn't part of this PR, but can we please tweak it to:

To expose the WCF service, change the configuration as shown below:

File: nservicebus/wcf/index.md [hunk 17]
```diff
@@ -46,14 +67,21 @@ And finally expose the WCF service via the config file, for the example above th
...
+partial:wcfcancellation
 
+partial:wcfrouting
 
 ## Queries and other return values
 
 To allow clients to perform queries, it is best not to use NServiceBus. Messaging is designed for non-blocking operations, and queries are (for the most part) operations for which the user wishes to wait.
 
-If there is some other operation that isn't strictly a query that returns a value for example some type of calculation consider invoking the operation locally where possible by referencing the DLL on the client.
+If there is some other operation that isn't strictly a query that returns a value, for example, some type of calculation consider invoking the operation locally where possible by referencing the DLL on the client.
```
Comment (indualagarsamy) [inline nservicebus/wcf/index.md]: >If there is some other operation that isn't strictly a query that returns a value

How about:

When performing operations that aren't straightforward as a simple query to return a value, for example, a long calculation, consider invoking the operation locally where possible by referencing the DLL on the client.

**Comment by indualagarsamy** [comment]
@danielmarbach - also currently the build is failing due to: `Could not find page reference for /nservicebus/wcf`

File: Snippets/Wcf/Wcf_1/Enum/Usage.cs (hunk 7 above)
Comment (SimonCropp) [inline Snippets/Wcf/Wcf_1/Enum/Usage.cs]: @SzymonPobiega leading and trailing whitespace in snippets is trimmed by the engine

File: Snippets/Wcf/Wcf_1/UpgradeGuides/5to6/Callbacks.cs [hunk 18]
```diff
@@ -0,0 +1,39 @@
+﻿namespace Wcf1
+{
+    using System.Configuration;
+    using System.Threading.Tasks;
+    using NServiceBus;
+
+    class Callbacks
```
Comment (danielmarbach) [inline Snippets/Wcf/Wcf_1/UpgradeGuides/5to6/Callbacks.cs]: @SimonCropp ups realized this is a left over which should be deleted

**Comment by danielmarbach** [comment]
Many thanks to @indualagarsamy @SimonCropp and @SzymonPobiega

**Pull request merged**